### PR TITLE
add refresh credentials property to loadTableResult

### PR DESCRIPTION
**Pull request by jasonf20**
Rebased version of https://github.com/apache/polaris/pull/1164

### CODE REVIEW

File: runtime/service/src/main/java/org/apache/polaris/service/catalog/iceberg/IcebergCatalogAdapter.java [hunk 1]
```diff
@@ -420,16 +423,45 @@ public Response loadTable(
...
         });
   }
 
+  private LoadTableResponse injectRefreshVendedCredentialProperties(
+      LoadTableResponse originalResponse, String credentialsEndpoint) {
+    LoadTableResponse.Builder loadResponseBuilder =
+        LoadTableResponse.builder().withTableMetadata(originalResponse.tableMetadata());
+    loadResponseBuilder.addAllConfig(originalResponse.config());
+    loadResponseBuilder.addAllCredentials(originalResponse.credentials());
+    loadResponseBuilder.addConfig(
```
Comment (singhpk234) [inline runtime/service/src/main/java/org/apache/polaris/service/catalog/iceberg/IcebergCatalogAdapter.java]: what about azure ? do we want to do it in follow-up ? also should we only inject this when creds are for aws ?

File: runtime/service/src/main/java/org/apache/polaris/service/catalog/iceberg/IcebergCatalogAdapter.java [hunk 2]
```diff
@@ -420,16 +423,45 @@ public Response loadTable(
...
           }
 
           return tryInsertETagHeader(Response.ok(response), response, namespace, table).build();
         });
   }
 
+  private LoadTableResponse injectRefreshVendedCredentialProperties(
+      LoadTableResponse originalResponse, String credentialsEndpoint) {
+    LoadTableResponse.Builder loadResponseBuilder =
+        LoadTableResponse.builder().withTableMetadata(originalResponse.tableMetadata());
```
Comment (singhpk234) [inline runtime/service/src/main/java/org/apache/polaris/service/catalog/iceberg/IcebergCatalogAdapter.java]: do we need to update metadata-location too ?

File: runtime/service/src/main/java/org/apache/polaris/service/catalog/iceberg/IcebergCatalogAdapter.java (hunk 1 above)
Comment (smaheshwar-pltr) [reply runtime/service/src/main/java/org/apache/polaris/service/catalog/iceberg/IcebergCatalogAdapter.java]: FWIW, I think enabling the Azure refresh properties currently causes Azure credential vending to immediately fail due to https://github.com/apache/iceberg/issues/13733, so could maybe wait for that fix to release?

Comment (smaheshwar-pltr) [reply runtime/service/src/main/java/org/apache/polaris/service/catalog/iceberg/IcebergCatalogAdapter.java]: > also should we only inject this when creds are for aws?

Could maybe lift the AWS refresh check below to `loadTable` itself, to avoid e.g. reconstructing the `LoadTableResponse` if not needed?

Comment (jasonf20) [reply runtime/service/src/main/java/org/apache/polaris/service/catalog/iceberg/IcebergCatalogAdapter.java]: I think it's better to provide the URL even if it's not AWS that way the user can override `AwsClientProperties.REFRESH_CREDENTIALS_ENABLED` if he is working with a system that he knows supports it despite the server not being updated to reflect this yet. For example, an updated Azure iceberg SDK like mentioned above.

File: runtime/service/src/main/java/org/apache/polaris/service/catalog/iceberg/IcebergCatalogAdapter.java (hunk 2 above)
Comment (jasonf20) [reply runtime/service/src/main/java/org/apache/polaris/service/catalog/iceberg/IcebergCatalogAdapter.java]: Update it to what value?

Do you mean is it redundant with the copyConfigs below? Seems like the metadata location is stored in a dedicated field and is not part of the config.

Comment (singhpk234) [reply runtime/service/src/main/java/org/apache/polaris/service/catalog/iceberg/IcebergCatalogAdapter.java]: nvm, it seems like the withTableMetadata sets both the tableMetada and the metadatLocation 
```
        public Builder withTableMetadata(TableMetadata tableMetadata) {
            this.metadataLocation = tableMetadata.metadataFileLocation();
            this.metadata = tableMetadata;
            return this;
        }
 ```
so it would not be required

File: runtime/service/src/main/java/org/apache/polaris/service/catalog/iceberg/IcebergCatalogAdapter.java (hunk 1 above)
Comment (singhpk234) [reply runtime/service/src/main/java/org/apache/polaris/service/catalog/iceberg/IcebergCatalogAdapter.java]: AWS cred refresh properties are : 

```
client.refresh-credentials-enabled
client.refresh-credentials-endpoint
```

Azure credential refresh properties are : 

```
adls.refresh-credentials-enabled
adls.refresh-credentials-endpoint
```

I am not sure how can one re-use aws cred endpoint prop to enable azure ? nevertheless why would server send back azure properties for an aws storage ?

Comment (jasonf20) [reply runtime/service/src/main/java/org/apache/polaris/service/catalog/iceberg/IcebergCatalogAdapter.java]: Oh, good to know. I assumed it's the same properties. I think it's not a good choice to use different properties, even if the client level has different settings. The client/server protocol should be standard. However, it seems like that boat has sailed. 

I would prefer having the endpoint URL always served even if it's not used, that way power users have the info available to them.

Despite that I've added a commit that only performs the inject code if the type is AWS. Let me know if that's better.

File: runtime/service/src/main/java/org/apache/polaris/service/catalog/iceberg/IcebergCatalogAdapter.java [hunk 3]
```diff
@@ -430,16 +433,48 @@ public Response loadTable(
...
+  private LoadTableResponse injectRefreshVendedCredentialProperties(
+      LoadTableResponse originalResponse, String credentialsEndpoint) {
+    // Only enable credential refresh for currently supported credential types
+    if (originalResponse.credentials().stream()
+        .anyMatch(
+            credential ->
+                credential
+                    .prefix()
+                    .toLowerCase()
+                    .startsWith(StorageConfigInfo.StorageTypeEnum.S3.name().toLowerCase()))) {
```
Comment (singhpk234) [inline runtime/service/src/main/java/org/apache/polaris/service/catalog/iceberg/IcebergCatalogAdapter.java]: i would also recommend adding this for azure, even if the client has bug, we would be ready from the server side.

File: runtime/service/src/main/java/org/apache/polaris/service/catalog/iceberg/IcebergCatalogAdapter.java (hunk 1 above)
Comment (singhpk234) [reply runtime/service/src/main/java/org/apache/polaris/service/catalog/iceberg/IcebergCatalogAdapter.java]: That i agree, unfortunately I not part of the discussions where these decisions were being made.
My recommendation is to also add similar handling for Azure, i understand the client has a bug with Azure, but this should not stop us at the server end to be ready with the changes.

**Comment by dimas-b** [comment]
Please address CI failures :)

File: runtime/service/src/main/java/org/apache/polaris/service/catalog/iceberg/IcebergCatalogAdapter.java (hunk 1 above)
Comment (jasonf20) [reply runtime/service/src/main/java/org/apache/polaris/service/catalog/iceberg/IcebergCatalogAdapter.java]: Alright, but since the properties are different for each provider, doing this in a clean way requires delegating this to the storage integration, so the code change scope becomes much larger. Please see the latest commit which does this.

File: polaris-core/src/main/java/org/apache/polaris/core/rest/PolarisResourcePaths.java [hunk 4]
```diff
@@ -57,6 +57,17 @@ public String genericTables(Namespace ns) {
...
 
+  public String credentialsPath(TableIdentifier ident) {
+    return SLASH.join(
+        "v1",
+        prefix,
+        "namespaces",
+        RESTUtil.encodeNamespace(ident.namespace()),
+        "tables",
+        RESTUtil.encodeString(ident.name()),
+        "credentials");
```
Comment (dimas-b) [inline polaris-core/src/main/java/org/apache/polaris/core/rest/PolarisResourcePaths.java:68]: Just to confirm: relative URIs are correct in this case, right?

File: runtime/service/src/main/java/org/apache/polaris/service/catalog/iceberg/IcebergCatalogAdapter.java [hunk 5]
```diff
@@ -430,9 +438,14 @@ public Response loadTable(
                     .loadTableIfStale(tableIdentifier, ifNoneMatch, snapshots)
                     .orElseThrow(() -> new WebApplicationException(Response.Status.NOT_MODIFIED));
           } else {
+            String refreshCredentialsEndpoint =
+                (delegationModes.contains(VENDED_CREDENTIALS))
+                    ? new PolarisResourcePaths(prefix).credentialsPath(tableIdentifier)
```
Comment (dimas-b) [inline runtime/service/src/main/java/org/apache/polaris/service/catalog/iceberg/IcebergCatalogAdapter.java]: Might be worth extracting into a shared method (cf. line 366).

File: polaris-core/src/main/java/org/apache/polaris/core/storage/cache/StorageCredentialCache.java [hunk 6]
```diff
@@ -130,7 +131,8 @@ public AccessConfig getOrGenerateSubScopeCreds(
                   polarisEntity.getType(),
                   k.allowedListAction(),
                   k.allowedReadLocations(),
-                  k.allowedWriteLocations());
+                  k.allowedWriteLocations(),
+                  refreshCredentialsEndpoint);
```
Comment (dimas-b) [inline polaris-core/src/main/java/org/apache/polaris/core/storage/cache/StorageCredentialCache.java]: `refreshCredentialsEndpoint` should be part of the cache key because `ScopedCredentialsResult` is a function also of `refreshCredentialsEndpoint`.

File: polaris-core/src/main/java/org/apache/polaris/core/storage/gcp/GcpCredentialsStorageIntegration.java [hunk 7]
```diff
@@ -75,7 +75,8 @@ public AccessConfig getSubscopedCreds(
       @Nonnull RealmConfig realmConfig,
       boolean allowListOperation,
       @Nonnull Set<String> allowedReadLocations,
-      @Nonnull Set<String> allowedWriteLocations) {
+      @Nonnull Set<String> allowedWriteLocations,
+      String refreshCredentialsEndpoint) {
```
Comment (dimas-b) [inline polaris-core/src/main/java/org/apache/polaris/core/storage/gcp/GcpCredentialsStorageIntegration.java]: Why is `refreshCredentialsEndpoint` used for AWS and Azure but not for GCP?

File: polaris-core/src/main/java/org/apache/polaris/core/storage/StorageAccessProperty.java [hunk 8]
```diff
@@ -39,6 +41,18 @@ public enum StorageAccessProperty {
       Boolean.class, "s3.path-style-access", "whether to use S3 path style access", false),
   CLIENT_REGION(
       String.class, "client.region", "region to configure client for making requests to AWS"),
+  AWS_REFRESH_CREDENTIALS_ENABLED(
+      Boolean.class,
+      AwsClientProperties.REFRESH_CREDENTIALS_ENABLED,
+      "whether to enable automatic refresh of credentials",
+      true,
```
Comment (dimas-b) [inline polaris-core/src/main/java/org/apache/polaris/core/storage/StorageAccessProperty.java]: I do not think this property is a credential property (logically)

File: polaris-core/src/main/java/org/apache/polaris/core/storage/StorageAccessProperty.java [hunk 9]
```diff
@@ -39,6 +41,18 @@ public enum StorageAccessProperty {
...
+      false),
+  AWS_REFRESH_CREDENTIALS_ENDPOINT(
+      String.class,
+      AwsClientProperties.REFRESH_CREDENTIALS_ENDPOINT,
+      "the endpoint to use for refreshing credentials",
+      true,
```
Comment (dimas-b) [inline polaris-core/src/main/java/org/apache/polaris/core/storage/StorageAccessProperty.java]: I do not think this property is a credential property (logically)

File: polaris-core/src/main/java/org/apache/polaris/core/storage/StorageAccessProperty.java [hunk 10]
```diff
@@ -52,6 +66,18 @@ public enum StorageAccessProperty {
   // it expects for SAS
   AZURE_ACCESS_TOKEN(String.class, "", "the azure scoped access token"),
   AZURE_SAS_TOKEN(String.class, "adls.sas-token.", "an azure shared access signature token"),
+  AZURE_REFRESH_CREDENTIALS_ENABLED(
+      Boolean.class,
+      "adls.refresh-credentials-enabled",
+      "whether to enable automatic refresh of credentials",
+      true,
```
Comment (dimas-b) [inline polaris-core/src/main/java/org/apache/polaris/core/storage/StorageAccessProperty.java]: I do not think this property is a credential property (logically)

File: polaris-core/src/main/java/org/apache/polaris/core/storage/StorageAccessProperty.java [hunk 11]
```diff
@@ -52,6 +66,18 @@ public enum StorageAccessProperty {
...
+      false),
+  AZURE_REFRESH_CREDENTIALS_ENDPOINT(
+      String.class,
+      "adls.refresh-credentials-endpoint",
+      "the endpoint to use for refreshing credentials",
+      true,
```
Comment (dimas-b) [inline polaris-core/src/main/java/org/apache/polaris/core/storage/StorageAccessProperty.java]: I do not think this property is a credential property (logically)

File: polaris-core/src/main/java/org/apache/polaris/core/storage/PolarisStorageIntegration.java [hunk 12]
```diff
@@ -61,7 +61,8 @@ public abstract AccessConfig getSubscopedCreds(
       @Nonnull RealmConfig realmConfig,
       boolean allowListOperation,
       @Nonnull Set<String> allowedReadLocations,
-      @Nonnull Set<String> allowedWriteLocations);
+      @Nonnull Set<String> allowedWriteLocations,
+      String refreshCredentialsEndpoint);
```
Comment (dimas-b) [inline polaris-core/src/main/java/org/apache/polaris/core/storage/PolarisStorageIntegration.java]: Maybe `Optional<String>`?

File: polaris-core/src/test/java/org/apache/polaris/core/storage/azure/AzureCredentialsStorageIntegrationTest.java [hunk 13]
```diff
@@ -32,20 +32,20 @@ public class AzureCredentialsStorageIntegrationTest {
   public void testAzureCredentialFormatting() {
     Instant expiresAt = Instant.ofEpochMilli(Long.MAX_VALUE);
 
-    AccessConfig noSuffixResult = toAccessConfig("sasToken", "some_account", expiresAt);
+    AccessConfig noSuffixResult = toAccessConfig("sasToken", "some_account", expiresAt, null);
```
Comment (dimas-b) [inline polaris-core/src/test/java/org/apache/polaris/core/storage/azure/AzureCredentialsStorageIntegrationTest.java]: Could you add tests for non-null refresh endpoint values? (sorry, if I missed them).

**Review by dimas-b — COMMENTED** [review]
LGTM overall :+1: Thanks for working on this @jasonf20 !

Do we have existing integration tests for the credential refresh endpoint? If not, could you add a test in `RestCatalogMinIOSpecialIT` (probably). I suppose we could using the test REST Client to validate refresh credential responses. I know this is specific to the S3 storage integration, but it will hit the common call path and we can validate Azure `AccessConfig` in unit tests. WDYT?

File: polaris-core/src/main/java/org/apache/polaris/core/rest/PolarisResourcePaths.java (hunk 4 above)
Comment (jasonf20) [reply polaris-core/src/main/java/org/apache/polaris/core/rest/PolarisResourcePaths.java:68]: Yes, from what I understand the recent version of the client automatically appends the base URI if a relative path is returned from the server.

File: polaris-core/src/main/java/org/apache/polaris/core/storage/cache/StorageCredentialCache.java (hunk 6 above)
Comment (jasonf20) [reply polaris-core/src/main/java/org/apache/polaris/core/storage/cache/StorageCredentialCache.java:135]: Since it's calculated statically I think it won't make a difference but I can add it. Perhaps if it's called with vended-credentials enabled and then again disabled it has some effect. Will add to the key.

File: polaris-core/src/main/java/org/apache/polaris/core/storage/gcp/GcpCredentialsStorageIntegration.java (hunk 7 above)
Comment (jasonf20) [reply polaris-core/src/main/java/org/apache/polaris/core/storage/gcp/GcpCredentialsStorageIntegration.java]: If it's supported in GCP I can use it, but previous comments mentioned it's only supported for AWS and Azure right now.

File: polaris-core/src/test/java/org/apache/polaris/core/storage/azure/AzureCredentialsStorageIntegrationTest.java (hunk 13 above)
Comment (jasonf20) [reply polaris-core/src/test/java/org/apache/polaris/core/storage/azure/AzureCredentialsStorageIntegrationTest.java]: I changed one of the tests to use a non null value already, I'll see if there are more places worth doing it.

File: polaris-core/src/main/java/org/apache/polaris/core/storage/StorageAccessProperty.java (hunk 8 above)
Comment (jasonf20) [reply polaris-core/src/main/java/org/apache/polaris/core/storage/StorageAccessProperty.java]: From my understanding of the client code:
1. The `AwsClientFactory` is initialized ahead of time with generic properties retrieved from the "configuration" endpoint. This doesn't include specific endpoints for credential refresh per table and/or access keys. 
2. Later on when loading the fileIO per table, the base properties + additional table properties provided from the `List<Credential>` are [added here](https://github.com/apache/iceberg/blob/2012f661a19e613abec2ef0024d7b340c9620f14/core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java#L450).

For the property to be available in `Credential.config()` down the line this needs to be set to true.

File: polaris-core/src/main/java/org/apache/polaris/core/storage/PolarisStorageIntegration.java (hunk 12 above)
Comment (jasonf20) [reply polaris-core/src/main/java/org/apache/polaris/core/storage/PolarisStorageIntegration.java:65]: done

**Comment by jasonf20** [comment]
@dimas-b I've added the integration test and unit test for azure as you suggested and made other fixes based on your comments.

File: polaris-core/src/main/java/org/apache/polaris/core/storage/gcp/GcpCredentialsStorageIntegration.java (hunk 7 above)
Comment (dimas-b) [reply polaris-core/src/main/java/org/apache/polaris/core/storage/gcp/GcpCredentialsStorageIntegration.java]: Interesting... but ok, let's implement what we can :)

File: polaris-core/src/main/java/org/apache/polaris/core/storage/StorageAccessProperty.java (hunk 8 above)
Comment (dimas-b) [reply polaris-core/src/main/java/org/apache/polaris/core/storage/StorageAccessProperty.java]: Thanks for the code pointer! In this case, it does make sense to mark these properties as "credentials", but please add a code comment for future readers to be aware of what Iceberg java clients expect.

File: polaris-core/src/main/java/org/apache/polaris/core/storage/PolarisCredentialVendor.java [hunk 14]
```diff
@@ -47,5 +48,6 @@ ScopedCredentialsResult getSubscopedCredsForEntity(
       PolarisEntityType entityType,
       boolean allowListOperation,
       @Nonnull Set<String> allowedReadLocations,
-      @Nonnull Set<String> allowedWriteLocations);
+      @Nonnull Set<String> allowedWriteLocations,
+      Optional<String> refreshCredentialsEndpoint);
```
Comment (dimas-b) [inline polaris-core/src/main/java/org/apache/polaris/core/storage/PolarisCredentialVendor.java:56]: Please add javadoc for this property and mention that it may be a relative URI.

File: polaris-core/src/main/java/org/apache/polaris/core/storage/PolarisStorageIntegration.java [hunk 15]
```diff
@@ -61,7 +62,8 @@ public abstract AccessConfig getSubscopedCreds(
       @Nonnull RealmConfig realmConfig,
       boolean allowListOperation,
       @Nonnull Set<String> allowedReadLocations,
-      @Nonnull Set<String> allowedWriteLocations);
+      @Nonnull Set<String> allowedWriteLocations,
+      Optional<String> refreshCredentialsEndpoint);
```
Comment (dimas-b) [inline polaris-core/src/main/java/org/apache/polaris/core/storage/PolarisStorageIntegration.java:70]: Same here: please add javadoc

File: polaris-core/src/main/java/org/apache/polaris/core/storage/cache/StorageCredentialCache.java (hunk 6 above)
Comment (dimas-b) [reply polaris-core/src/main/java/org/apache/polaris/core/storage/cache/StorageCredentialCache.java]: Thx for adding it to the cache key! Would you mind accessing it here via `k.` for clarity?

File: runtime/service/src/main/java/org/apache/polaris/service/catalog/iceberg/IcebergCatalogAdapter.java [hunk 16]
```diff
@@ -430,16 +438,28 @@ public Response loadTable(
...
 
           return tryInsertETagHeader(Response.ok(response), response, namespace, table).build();
         });
   }
 
+  private static Optional<String> getRefreshCredentialsEndpoint(
+      EnumSet<AccessDelegationMode> delegationModes,
+      String prefix,
+      TableIdentifier tableIdentifier) {
+    return delegationModes.contains(AccessDelegationMode.VENDED_CREDENTIALS)
```
Comment (dimas-b) [inline runtime/service/src/main/java/org/apache/polaris/service/catalog/iceberg/IcebergCatalogAdapter.java:458]: This logic is ok from my POV, but I wonder if other community members prefer a feature flag to switch the new behaviour on/off.

Please send an email about this to the `dev` ML.

**Review by dimas-b — COMMENTED** [review]
This PR LGTM in its current state :+1: ... with some minor comments :slightly_smiling_face: 

However, I believe it is worth sending a `dev` email about the new feature before merging (as commented below).

File: runtime/service/src/main/java/org/apache/polaris/service/catalog/iceberg/IcebergCatalogAdapter.java (hunk 16 above)
Comment (dimas-b) [reply runtime/service/src/main/java/org/apache/polaris/service/catalog/iceberg/IcebergCatalogAdapter.java:458]: Actually, I'll send the `dev` email myself :)

Comment (dimas-b) [reply runtime/service/src/main/java/org/apache/polaris/service/catalog/iceberg/IcebergCatalogAdapter.java:458]: https://lists.apache.org/thread/dvrx5lg36ovo7387pt6dm06wgmswstvc

File: polaris-core/src/main/java/org/apache/polaris/core/storage/StorageAccessProperty.java [hunk 17]
```diff
@@ -39,6 +41,18 @@ public enum StorageAccessProperty {
...
+  AWS_REFRESH_CREDENTIALS_ENABLED(
+      Boolean.class,
+      AwsClientProperties.REFRESH_CREDENTIALS_ENABLED,
+      "whether to enable automatic refresh of credentials",
+      true,
+      false),
+  AWS_REFRESH_CREDENTIALS_ENDPOINT(
+      String.class,
+      AwsClientProperties.REFRESH_CREDENTIALS_ENDPOINT,
+      "the endpoint to use for refreshing credentials",
```
Comment (singhpk234) [inline polaris-core/src/main/java/org/apache/polaris/core/storage/StorageAccessProperty.java]: ideally we are not refreshing the credentials, we are getting brand new one how about we say something like 
https://github.com/apache/iceberg/blob/main/open-api/rest-catalog-open-api.yaml#L119

**Review by singhpk234 — COMMENTED** [review]
Mostly LGTM as well 👍 ! we are very close :) !

**Comment by dimas-b** [comment]
@jasonf20 : Could you add an entry about this change to `CHANGELOG.md`, section `New Features`?

File: polaris-core/src/main/java/org/apache/polaris/core/storage/gcp/GcpCredentialsStorageIntegration.java (hunk 7 above)
Comment (dimas-b) [reply polaris-core/src/main/java/org/apache/polaris/core/storage/gcp/GcpCredentialsStorageIntegration.java]: It looks like GCP credential refresh _is_ supported: https://github.com/lakekeeper/lakekeeper/pull/1227/files#diff-8d3c6564c330f473c1e4556b7dbbf79dddae3a49a9d9c55352c7a614fc8b3fb5R88 WDYT?

Comment (dimas-b) [reply polaris-core/src/main/java/org/apache/polaris/core/storage/gcp/GcpCredentialsStorageIntegration.java]: Cf. https://github.com/apache/iceberg/blob/b82dac4858fb2d15929a797660252c50f8e00ed8/gcp/src/main/java/org/apache/iceberg/gcp/GCPProperties.java#L50

Comment (dimas-b) [reply polaris-core/src/main/java/org/apache/polaris/core/storage/gcp/GcpCredentialsStorageIntegration.java]: Iceberg support for GCP credential refresh appears to be available since 1.7.0, according to https://github.com/apache/iceberg/pull/11282

File: polaris-core/src/main/java/org/apache/polaris/core/storage/StorageAccessProperty.java (hunk 8 above)
Comment (jasonf20) [reply polaris-core/src/main/java/org/apache/polaris/core/storage/StorageAccessProperty.java]: I had another look at this. I think I was wrong since it also passes in `tableConf` which if set initializes a new fileIO with those properties, so this can be false. I'll update those.

File: polaris-core/src/main/java/org/apache/polaris/core/storage/gcp/GcpCredentialsStorageIntegration.java (hunk 7 above)
Comment (jasonf20) [reply polaris-core/src/main/java/org/apache/polaris/core/storage/gcp/GcpCredentialsStorageIntegration.java]: done

File: polaris-core/src/main/java/org/apache/polaris/core/storage/PolarisCredentialVendor.java [hunk 18]
```diff
@@ -37,6 +38,10 @@ public interface PolarisCredentialVendor {
    *     allowedWriteLocations
    * @param allowedReadLocations a set of allowed to read locations
    * @param allowedWriteLocations a set of allowed to write locations
+   * @param refreshCredentialsEndpoint an optional endpoint to use for refreshing credentials. If
+   *     supported by the storage type it will be returned to the client in the appropriate
+   *     properties. The endpoint may be relative to the base URI and the client is responsible for
```
Comment (singhpk234) [inline polaris-core/src/main/java/org/apache/polaris/core/storage/PolarisCredentialVendor.java:43]: can we return anything except relative path ? my understanding is base uri is always catalog uri https://github.com/apache/iceberg/blob/main/aws/src/main/java/org/apache/iceberg/aws/s3/VendedCredentialsProvider.java#L92  is this statement in intentionally open ended since spec is not explicit ?

**Review by dimas-b — DISMISSED** [review]
LGTM :+1: Thanks again for you contribution, @jasonf20 !

Since this PR was already mentioned on the [dev ML](https://lists.apache.org/thread/dvrx5lg36ovo7387pt6dm06wgmswstvc) on Aug 21 and no concerned were raised, I think it is good to merge. I'll wait until Aug 26 before merging, in case any last minute comments come up.

File: CHANGELOG.md [hunk 19]
```diff
@@ -81,6 +81,9 @@ at locations that better optimize for object storage.
 
 - Introduced bootstrap command options to specify custom schema files for database initialization.
 
+- Added refresh credentials endpoint configuration to LoadTableResponse for AWS, Azure, and GCP. Enabling
+automatic credential refresh per table on the client side.
```
Comment (singhpk234) [inline CHANGELOG.md]: nit: I understand the intent but i think imho its better to be explicit here, as there.
```suggestion
automatic storage credential refresh per table on the client side.
```

File: polaris-core/src/main/java/org/apache/polaris/core/storage/gcp/GcpCredentialsStorageIntegration.java (hunk 7 above)
Comment (dimas-b) [reply polaris-core/src/main/java/org/apache/polaris/core/storage/gcp/GcpCredentialsStorageIntegration.java]: thx !

**Review by singhpk234 — APPROVED** [review]
LGTM overall (have some small cosmetic suggestion for documentation), Thank you @jasonf20 ! much needed feature for long running jobs !

will wait for @dimas-b review and approval  before merge

File: polaris-core/src/main/java/org/apache/polaris/core/storage/StorageAccessProperty.java (hunk 8 above)
Comment (dimas-b) [reply polaris-core/src/main/java/org/apache/polaris/core/storage/StorageAccessProperty.java]: Thanks for your diligence! :slightly_smiling_face:

**Comment by singhpk234** [comment]
Looks like we already have @dimas-b ;) review, we were parallely reviewing, we good to go from my end !

File: CHANGELOG.md (hunk 19 above)
Comment (dimas-b) [reply CHANGELOG.md]: The suggestion LGTM :+1:

**Comment by dimas-b** [comment]
Oops, CI failure :sweat_smile:

File: polaris-core/src/main/java/org/apache/polaris/core/storage/PolarisCredentialVendor.java (hunk 18 above)
Comment (jasonf20) [reply polaris-core/src/main/java/org/apache/polaris/core/storage/PolarisCredentialVendor.java:43]: At one point the client required full paths, now it works with partial paths. I'm not sure if the spec identifies one as the correct method over the other so I left it like this. If one is definitely the "correct" way I can update the comment.

Comment (dimas-b) [reply polaris-core/src/main/java/org/apache/polaris/core/storage/PolarisCredentialVendor.java:43]: @jasonf20 : do you know the Iceberg version / PR where relative paths started working?

Comment (jasonf20) [reply polaris-core/src/main/java/org/apache/polaris/core/storage/PolarisCredentialVendor.java:43]: It was linked in the previous PR. https://github.com/apache/polaris/pull/1164#discussion_r2007984981

Seems like it's since 1.8.0

**Comment by jasonf20** [comment]
> Oops, CI failure 😅

fixed

File: polaris-core/src/main/java/org/apache/polaris/core/storage/PolarisCredentialVendor.java (hunk 18 above)
Comment (singhpk234) [reply polaris-core/src/main/java/org/apache/polaris/core/storage/PolarisCredentialVendor.java:43]: Oh no this made a lot of things super tricky, i checked iceberg python for refresh keyword but i didn't any prs for that so i am assuming that it doesn't support, taking only the iceberg java sdk reference 
if we always send relative url it only works for >= 1.8
fails for < 1.8 
I know there is a header that the java sdk sends which has SDK version `X-Client-Version` header we may want to salvage this now on when to return the absolute vs relative, https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/rest/HTTPClient.java#L77

seems like the only broken release would be 1.7 

Its a bit surprising to me, but i think for now we can park this by just documenting this gotcha, WDYT @dimas-b ?

Comment (dimas-b) [reply polaris-core/src/main/java/org/apache/polaris/core/storage/PolarisCredentialVendor.java:43]: Getting full URI is a bit trickier - as discussed in #1164.

From my POV, let's merge this "as is" and open a GH issue that it does not work properly for Iceberg < 1.8. When someone has time, let's improve.

Getting this feature enabled for Iceberg clients >= 1.8 is valuable, I think.

Comment (dimas-b) [reply polaris-core/src/main/java/org/apache/polaris/core/storage/PolarisCredentialVendor.java:43]: Since old client version behave differently, I think we ought to have a feature flag now... just in case someone runs into client incompatibility issues and wants to disable credential refresh properties completely (i.e. revert to old Polaris behaviour). WDYT?

Comment (singhpk234) [reply polaris-core/src/main/java/org/apache/polaris/core/storage/PolarisCredentialVendor.java:43]: That would be ideal IMHO (specially now), but i think the jobs without us sending the credentials endpoint would still have failed because my understanding is we would have not made the /credentials call unless the creds in loadTable expires https://github.com/apache/iceberg/blob/main/aws/src/main/java/org/apache/iceberg/aws/s3/VendedCredentialsProvider.java#L63, its just we might see an HTTP error rather than the creds expired exception.
But agree, we might want to have a feature flag to fallback to the old behaviour if we unexpectedly fail.

Comment (jasonf20) [reply polaris-core/src/main/java/org/apache/polaris/core/storage/PolarisCredentialVendor.java:43]: I don't think a feature flag is needed. 

1. The client is already broken, if they reach the need to refresh credentials it already fails today. 
2. A feature flag will not solve the problem. Enabling it will fix the issue for some and create a different issue for others, so you'll never be able to enable it anymore than you can without the flag. 
3. Clients can simply upgrade to 1.8.0 to get a working version. Their current version doesn't work anyway so they can upgrade at their own time when this becomes important to them. 
4. If you would like we can remove the `refresh_credentials_enabled` property and let the client set that one in his catalog settings. That way he can control if this gets used or not. This would be better than a flag and is controlled client side. not server side.

Comment (jasonf20) [reply polaris-core/src/main/java/org/apache/polaris/core/storage/PolarisCredentialVendor.java:43]: I've added a commit that does 4.

**Review by dimas-b — COMMENTED** [review]
@jasonf20 : I'm a bit lost now... Why do we need to remove the "enabled" flag in https://github.com/apache/polaris/pull/2341/commits/2a123d1df204c748622ff4b85f3a8906daaa8a9c?

In my earlier comments I mean that we need to add a new flag to `FeatureConfiguration` and use it to control whether to add the credential refresh endpoint and the related "enabled" flag to `AccessConfig`... WDYT?

**Review by dimas-b — DISMISSED** [review]
After re-reading @jasonf20 comment, I think removing of the "enabled" properties from `AccessConfig` is a sound approach.

File: CHANGELOG.md [hunk 20]
```diff
@@ -81,6 +81,9 @@ at locations that better optimize for object storage.
 
 - Introduced bootstrap command options to specify custom schema files for database initialization.
 
+- Added refresh credentials endpoint configuration to LoadTableResponse for AWS, Azure, and GCP. Enabling
+automatic storage credential refresh per table on the client side.
```
Comment (singhpk234) [inline CHANGELOG.md]: Lets call out the client version limitation here and the suggestion to the caller to set the client side override to enable this.

**Review by singhpk234 — DISMISSED** [review]

File: CHANGELOG.md (hunk 20 above)
Comment (dimas-b) [reply CHANGELOG.md]: good points!

Comment (dimas-b) [reply CHANGELOG.md]: since the release is imminent, if the round trip cycle is slow, I suppose we can still merge this PR and later adjust the changelog between @singhpk234 and myself :) Let's give it a few more hours.

Comment (jasonf20) [reply CHANGELOG.md]: I updated the changelog. Seems a little verbose now for a changelog but It's probably fine. 

p.s. I believe you should be able to push changes to my branch yourself since I have the "Allow edits and access to secrets by maintainers" checkbox enabled. But I never tried using that myself.

File: CHANGELOG.md [hunk 21]
```diff
@@ -82,7 +82,9 @@ at locations that better optimize for object storage.
 - Introduced bootstrap command options to specify custom schema files for database initialization.
 
 - Added refresh credentials endpoint configuration to LoadTableResponse for AWS, Azure, and GCP. Enabling
-automatic storage credential refresh per table on the client side.
+automatic storage credential refresh per table on the client side. Java client version >= 1.8.0 is required.
+The endpoint path is always returned when using vended credentials, but the client must configure 
+REFRESH\_ENDPOINT\_ENABLED = "true" in the catalog properties for the client to use the endpoint.
```
Comment (singhpk234) [inline CHANGELOG.md]: the properties are unfortunately different for the all the cloud providers, not sure if we wanna call this out, just saying enabled never the less `refresh-credentials-enabled` should be endpoint 

how about we say that clients must enable refresh-credentials flag for their respective cloud provider ?

Comment (jasonf20) [reply CHANGELOG.md]: Yea that's less verbose. I changed it to that.

Comment (dimas-b) [reply CHANGELOG.md]: lgtm :+1:

**Comment by jasonf20** [comment]
Thanks for your reviews @dimas-b @singhpk234 !

File: CHANGELOG.md [hunk 22]
```diff
@@ -81,6 +81,11 @@ at locations that better optimize for object storage.
 
 - Introduced bootstrap command options to specify custom schema files for database initialization.
 
+- Added refresh credentials endpoint configuration to LoadTableResponse for AWS, Azure, and GCP. Enabling
+automatic storage credential refresh per table on the client side. Java client version >= 1.8.0 is required.
+The endpoint path is always returned when using vended credentials, but clients must enable the
```
Comment (flyrain) [inline CHANGELOG.md:86]: We will need to document this. Unfortunately, the best place to doc this is a page for storage setting and credential vending, which doesn't exist yet. Here is the issue filed long time ago, https://github.com/apache/polaris/issues/1325.  Not a blocker for this PR though.
